enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,249 @@
+- In the *live ISO* flow, it is not possible to notify Ironic (and thus
+  baremetal-operator) when the actual installation finishes. From the
+  perspective of Metal3 components, the installation is done once the ISO is
+  connected. As a result, Ironic cannot disconnect the ISO from the machine's
+  BMC, nor can it update the BMC's boot device settings properly.
+- The *live ISO* flow disables Ironic *inspection* process and has to simulate
+  it by updating the `BareMetalHost` based on the assisted service's inspection
+  process. This creates an unnessesary coupling between the assisted service
+  and the baremetal-operator's hardware inventory format (which, in turn, is
+  derived from the Ironic's inventory format).
+- The *standard* flow is highly [extensible][hardware managers] via the
+  [ironic-agent-image][ironic-agent] container image. This extensibility does
+  not apply to the *live ISO* flow since it does not use the Ironic agent.
+
+### Goals
+
+- ZTP follows the *standard* Metal3 flow instead of *live ISO*.
+- (i)PXE boot, firwmare and RAID settings work for ZTP without introducing
+  special API in Metal3 components.
+
+### Non-Goals
+
+- Converging two agent implementations (Ironic and assisted) into a single one.
+- Making assisted components mandatory for all bare-metal deployments.
+
+## Proposal
+
+- Building CoreOS images for the *standard* flow is handled by
+  [image-customization-controller][icc] (ICC). It will be updated to use the
+  assisted service's `InfraEnv` resources, when they are present for the given
+  `BareMetalHost`.
+- When `InfraEnv` resources are present, the generated images will contain
+  Ignition configuration for both agents. The Ironic agent will be started
+  automatically and handle coordination with Ironic and baremetal-operator.
+  The configuration will also supress automatic start-up of the assisted
+  agent.
+- [Ironic agent][ironic-agent] currently contains a custom deployment plugin
+  (*custom deploy method* in baremetal-operator terms, *custom deploy step* in
+  Ironic terms) that invokes `coreos-installer` for the image installation.
+  A new custom deploy method `install_assisted` will be added to start the
+  assisted agent at the right moment.
+- The assisted service will stop explicitly disabling *inspection* on
+  `BareMetalHost`s it manages, will start using a different process based
+  on the  new `install_assisted` deploy method and stop rebooting the machine
+  in the end of the process.
+
+### User Stories
+
+N/A (this change itself does not add new user stories)
+
+### API Extensions
+
+All required API extensions are already present.
+
+### Implementation Details/Notes/Constraints
+
+#### Image building
+
+ICC will follow different code paths depending on whether the `BareMetalHost`
+has the `infraenvs.agent-install.openshift.io` annotation linking it to an
+`InfraEnv` resource. In both cases it will generate an Ignition configuration
```

Comment:
Do we need some coordination here between ICC and BMAC when an InfraEnv is updated?

Reply:
> Do we need some coordination here between ICC and BMAC when an InfraEnv is updated?

Ideally, with this work, BMAC should be able to stop syncing the BMH with the `InfraEnv` and the `Agent`. This means, the agent inspection will only be used by AI (instead of it being propagated to the BMH). The InfraEnv URL will be read by ICC.

Finally, we will likely stop setting the `baremetalhost.metal3.io/detached` annotation. At the very least, we would set it if and only if the deployment succeeded but I don't think we will need to set this annotation anymore

Reply:
I don't think there's any connection between the `detached` annotation and this work.
Detaching the host means that ironic in the hub cluster won't continue to probe it (for power state) and metal³ won't attempt to manage it. You should do this whenever there is a BaremetalHost created in a spoke cluster that is self-managing the host, and probably not if there is not. Either way though, it is independent of this enhancement.